packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +424,411 @@ contract AmmPool is IBlockReceiver {
+    }
+
+    function processAmmUpdates(
+        Context memory ctx,
+        bool   start
+        )
+        internal
+    {
+        for (uint i = 0; i < ctx.tokens.length; i++) {
             // Check that the AMM update in the block matches the expected update
-            AmmUpdateTransaction.AmmUpdate memory update = _block.readAmmUpdate(txIdx++);
+            AmmUpdateTransaction.AmmUpdate memory update = ctx._block.readAmmUpdate(ctx.txIdx++);
             require(update.owner == address(this), "INVALID_TX_DATA");
             require(update.accountID == accountID, "INVALID_TX_DATA");
-            require(update.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
+            require(update.tokenID == ctx.tokens[i].tokenID, "INVALID_TX_DATA");
             require(update.feeBips == feeBips, "INVALID_TX_DATA");
-            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            require(update.tokenWeight == (start ? 0 : ctx.tokens[i].weight), "INVALID_TX_DATA");
             // Now approve this AMM update
             update.validUntil = 0xffffffff;
-            bytes32 txHash = AmmUpdateTransaction.hashTx(exchangeDomainSeparator, update);
+            bytes32 txHash = AmmUpdateTransaction.hashTx(ctx.exchangeDomainSeparator, update);
             exchange.approveTransaction(address(this), txHash);
-            numTransactionsConsumed++;
-            // AMM account balance now available onchain
-            ammBalancesInAccount[i] = update.balance;
-            ammBalances[i] = ammBalancesInAccount[i];
+            ctx.numTransactionsConsumed++;
+            if (start) {
+                // AMM account balance now available onchain
+                ctx.ammBalancesInAccount[i] = update.balance;
+                ctx.ammBalances[i] = ctx.ammBalancesInAccount[i];
+            } else {
+                require(ctx.ammBalances[i] == update.balance, "UNEXPECTED_AMM_BALANCE");
+            }
         }
+    }
 
-        // Process work in the queue
-        for (uint n = queuePos; n < queuePos + auxData.numItems; n++) {
-            QueueItem memory item = queue[n];
-            // TODO: all the necessary pool logic, mint/burn LP tokens,...
-            if (item.workType == WorkType.JOIN) {
-                PoolJoin memory join = item.join;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = join.maxAmountsIn[i];
-                    ammBalances[i] = ammBalances[i].add(amount);
-                    // Unlock the amount locked for this join
-                    locked[item.owner][tokens[i].addr] = locked[item.owner][tokens[i].addr].sub(join.maxAmountsIn[i]);
+    function processJoin(
+        Context  memory ctx,
+        PoolJoin memory join,
+        bytes    memory signature
+        )
+        internal
+    {
+        bytes32 poolTxHash = hashPoolJoin(ctx.DOMAIN_SEPARATOR, join);
+        if (signature.length == 0) {
```

Comment:
is this an ECDSA signature, if so, when a signature is provided, this can still be a layer-1 join just using a onchain signature verification. Asserting `require(join.fromLayer2, "INVALID_DATA")` will be incorrect.

We have:

- layer1-join with onchain hash approval
- layer1-join with ECDSA signature approval
- layer-2 join with transfers (EdDSA signature approval)